week-1-ts-ecosystem/w01e03-object-oriented-structure/models/Basket.ts

```diff
@@ -0,0 +1,119 @@
+import { Product, ProductType, UpdateProductDTO } from "./Product";
+
+type NonEmptyProductArray = [Product, ...Product[]];
+
+export class Basket {
+  name: string = "";
+  #products: Product[] = [];
+  type?: ProductType;
```

Comment:
W momencie w którym nie podzielisz sobie `Product` na odpowiednie klasy w zależności od `ProductType` i nie zastosujesz tutaj podejścia generycznego koszyka,
Twoja logika @wszczepanski97 multiplikuje się do potrzeb dopisywania:

`areProductsTypeConsistent`

`isProductTypeEqualBasketType`

Nota bene mogły by to być pola - referencje do funkcji, prywatne (`#` lub `private`).

To tak:
- [x] `ProductType` jest dobrym pomysłem
- [ ] `ProductType` powinien być dopisany do danej klasy produktu (a nie jako pole koszyka)
- [ ] `Basket` może przechowywać tylko klasę spełniającą dany `ProductType`

wtedy nie potrzebujesz tych wszystkich funkcji i koszyk "pilnuje się sam" - TypeScriptem.

Jesteś bardzo blisko żeby to uprościć.
Tylko wprowadź proszę odpowiednie dziedziczenie z klasy `Product` dla konkretnych produktów - i dodaj im odpowiednie wartości `ProductType`.